// astro/dist/vite-plugin-astro-server/plugin.js
import { AsyncLocalStorage } from "node:async_hooks";
import { IncomingMessage } from "node:http";
import { isRunnableDevEnvironment } from "vite";
import { toFallbackType } from "../core/app/common.js";
import { toRoutingStrategy } from "../core/app/entrypoints/index.js";
import { ASTRO_VITE_ENVIRONMENT_NAMES, devPrerenderMiddlewareSymbol } from "../core/constants.js";
import {
  getAlgorithm,
  getDirectives,
  getScriptHashes,
  getScriptResources,
  getStrictDynamic,
  getStyleHashes,
  getStyleResources,
  shouldTrackCspHashes
} from "../core/csp/common.js";
import { createKey, getEnvironmentKey, hasEnvironmentKey } from "../core/encryption.js";
import { getViteErrorPayload } from "../core/errors/dev/index.js";
import { AstroError, AstroErrorData } from "../core/errors/index.js";
import { NOOP_MIDDLEWARE_FN } from "../core/middleware/noop-middleware.js";
import { createViteLoader } from "../core/module-loader/index.js";
import { isRouteServerIsland, matchAllRoutes } from "../core/routing/match.js";
import { resolveMiddlewareMode } from "../integrations/adapter-utils.js";
import { SERIALIZED_MANIFEST_ID } from "../manifest/serialized.js";
import { ASTRO_DEV_SERVER_APP_ID } from "../vite-plugin-app/index.js";
import { baseMiddleware } from "./base.js";
import { createController } from "./controller.js";
import { recordServerError } from "./error.js";
import { setRouteError } from "./server-state.js";
import { routeGuardMiddleware } from "./route-guard.js";
import { secFetchMiddleware } from "./sec-fetch.js";
import { trailingSlashMiddleware } from "./trailing-slash.js";
import { sessionConfigToManifest } from "../core/session/utils.js";
function createVitePluginAstroServer({
  settings,
  logger
}) {
  return {
    name: "astro:server",
    applyToEnvironment(environment) {
      return environment.name === ASTRO_VITE_ENVIRONMENT_NAMES.ssr;
    },
    async configureServer(viteServer) {
      const ssrEnvironment = viteServer.environments[ASTRO_VITE_ENVIRONMENT_NAMES.ssr];
      const prerenderEnvironment = viteServer.environments[ASTRO_VITE_ENVIRONMENT_NAMES.prerender];
      const runnableSsrEnvironment = isRunnableDevEnvironment(ssrEnvironment) ? ssrEnvironment : void 0;
      const runnablePrerenderEnvironment = isRunnableDevEnvironment(prerenderEnvironment) ? prerenderEnvironment : void 0;
      if (!runnableSsrEnvironment && !runnablePrerenderEnvironment) {
        return;
      }
      async function createHandler(environment) {
        const loader = createViteLoader(viteServer, environment);
        const { default: createAstroServerApp } = await environment.runner.import(ASTRO_DEV_SERVER_APP_ID);
        const controller = createController({ loader });
        const { handler } = await createAstroServerApp(controller, settings, loader, logger);
        const { manifest } = await environment.runner.import(SERIALIZED_MANIFEST_ID);
        return { controller, handler, loader, manifest, environment };
      }
      const ssrHandler = runnableSsrEnvironment ? await createHandler(runnableSsrEnvironment) : void 0;
      const prerenderHandler = runnablePrerenderEnvironment ? await createHandler(runnablePrerenderEnvironment) : void 0;
      const localStorage = new AsyncLocalStorage();
      function handleUnhandledRejection(rejection) {
        const error = AstroError.is(rejection) ? rejection : new AstroError({
          ...AstroErrorData.UnhandledRejection,
          message: AstroErrorData.UnhandledRejection.message(rejection?.stack || rejection)
        });
        const store = localStorage.getStore();
        const handlers = [];
        if (ssrHandler) handlers.push(ssrHandler);
        if (prerenderHandler) handlers.push(prerenderHandler);
        for (const currentHandler of handlers) {
          if (store instanceof IncomingMessage) {
            setRouteError(currentHandler.controller.state, store.url, error);
          }
          const { errorWithMetadata } = recordServerError(
            currentHandler.loader,
            currentHandler.manifest,
            logger,
            error
          );
          setTimeout(
            async () => currentHandler.loader.webSocketSend(await getViteErrorPayload(errorWithMetadata)),
            200
          );
        }
      }
      process.on("unhandledRejection", handleUnhandledRejection);
      viteServer.httpServer?.on("close", () => {
        process.off("unhandledRejection", handleUnhandledRejection);
      });
      return () => {
        const shouldHandlePrerenderInCore = Boolean(
          viteServer[devPrerenderMiddlewareSymbol]
        );
        if (!ssrHandler && !(prerenderHandler && shouldHandlePrerenderInCore)) {
          return;
        }
        viteServer.middlewares.stack.unshift({
          route: "",
          handle: baseMiddleware(settings, logger)
        });
        viteServer.middlewares.stack.unshift({
          route: "",
          handle: trailingSlashMiddleware(settings)
        });
        viteServer.middlewares.stack.unshift({
          route: "",
          handle: routeGuardMiddleware(settings)
        });
        viteServer.middlewares.stack.unshift({
          route: "",
          handle: secFetchMiddleware(logger, settings.config.security?.allowedDomains)
        });
        if (prerenderHandler && shouldHandlePrerenderInCore) {
          viteServer.middlewares.use(
            async function astroDevPrerenderHandler(request, response, next) {
              if (request.url === void 0 || !request.method) {
                response.writeHead(500, "Incomplete request");
                response.end();
                return;
              }
              if (request.url.startsWith("/@") || request.url.startsWith("/__")) {
                return next();
              }
              if (request.url.includes("/node_modules/")) {
                return next();
              }
              try {
                const pathname = decodeURI(new URL(request.url, "http://localhost").pathname);
                const { routes } = await prerenderHandler.environment.runner.import("virtual:astro:routes");
                const routesList = { routes: routes.map((r) => r.routeData) };
                const matches = matchAllRoutes(pathname, routesList);
                if (!matches.some((route) => route.prerender || isRouteServerIsland(route))) {
                  return next();
                }
                localStorage.run(request, () => {
                  prerenderHandler.handler(request, response);
                });
              } catch (err) {
                next(err);
              }
            }
          );
        }
        if (ssrHandler) {
          viteServer.middlewares.use(async function astroDevHandler(request, response) {
            if (request.url === void 0 || !request.method) {
              response.writeHead(500, "Incomplete request");
              response.end();
              return;
            }
            localStorage.run(request, () => {
              ssrHandler.handler(request, response);
            });
          });
        }
      };
    }
  };
}
async function createDevelopmentManifest(settings) {
  let i18nManifest;
  let csp;
  if (settings.config.i18n) {
    i18nManifest = {
      fallback: settings.config.i18n.fallback,
      strategy: toRoutingStrategy(settings.config.i18n.routing, settings.config.i18n.domains),
      defaultLocale: settings.config.i18n.defaultLocale,
      locales: settings.config.i18n.locales,
      domainLookupTable: {},
      fallbackType: toFallbackType(settings.config.i18n.routing),
      domains: settings.config.i18n.domains
    };
  }
  if (shouldTrackCspHashes(settings.config.security.csp)) {
    const styleHashes = [
      ...getStyleHashes(settings.config.security.csp),
      ...settings.injectedCsp.styleHashes
    ];
    csp = {
      cspDestination: settings.adapter?.adapterFeatures?.staticHeaders ? "adapter" : void 0,
      scriptHashes: getScriptHashes(settings.config.security.csp),
      scriptResources: getScriptResources(settings.config.security.csp),
      styleHashes,
      styleResources: getStyleResources(settings.config.security.csp),
      algorithm: getAlgorithm(settings.config.security.csp),
      directives: getDirectives(settings),
      isStrictDynamic: getStrictDynamic(settings.config.security.csp)
    };
  }
  return {
    rootDir: settings.config.root,
    srcDir: settings.config.srcDir,
    cacheDir: settings.config.cacheDir,
    outDir: settings.config.outDir,
    buildServerDir: settings.config.build.server,
    buildClientDir: settings.config.build.client,
    publicDir: settings.config.publicDir,
    trailingSlash: settings.config.trailingSlash,
    buildFormat: settings.config.build.format,
    compressHTML: settings.config.compressHTML,
    assetsDir: settings.config.build.assets,
    serverLike: settings.buildOutput === "server",
    middlewareMode: resolveMiddlewareMode(settings.adapter?.adapterFeatures),
    assets: /* @__PURE__ */ new Set(),
    entryModules: {},
    routes: [],
    adapterName: settings?.adapter?.name ?? "",
    clientDirectives: settings.clientDirectives,
    renderers: [],
    base: settings.config.base,
    userAssetsBase: settings.config?.vite?.base,
    assetsPrefix: settings.config.build.assetsPrefix,
    site: settings.config.site,
    componentMetadata: /* @__PURE__ */ new Map(),
    inlinedScripts: /* @__PURE__ */ new Map(),
    i18n: i18nManifest,
    checkOrigin: (settings.config.security?.checkOrigin && settings.buildOutput === "server") ?? false,
    actionBodySizeLimit: settings.config.security?.actionBodySizeLimit ? settings.config.security.actionBodySizeLimit : 1024 * 1024,
    // 1mb default
    serverIslandBodySizeLimit: settings.config.security?.serverIslandBodySizeLimit ? settings.config.security.serverIslandBodySizeLimit : 1024 * 1024,
    // 1mb default
    key: hasEnvironmentKey() ? getEnvironmentKey() : createKey(),
    middleware() {
      return {
        onRequest: NOOP_MIDDLEWARE_FN
      };
    },
    sessionConfig: sessionConfigToManifest(settings.config.session),
    csp,
    image: {
      objectFit: settings.config.image.objectFit,
      objectPosition: settings.config.image.objectPosition,
      layout: settings.config.image.layout
    },
    devToolbar: {
      enabled: settings.config.devToolbar.enabled && await settings.preferences.get("devToolbar.enabled"),
      latestAstroVersion: settings.latestAstroVersion,
      debugInfoOutput: "",
      placement: settings.config.devToolbar.placement
    },
    logLevel: settings.logLevel,
    shouldInjectCspMetaTags: false,
    experimentalQueuedRendering: {
      enabled: !!settings.config.experimental?.queuedRendering,
      poolSize: settings.config.experimental?.queuedRendering?.poolSize ?? 1e3
    }
  };
}
export {
  createDevelopmentManifest,
  createVitePluginAstroServer as default
};
