// astro/dist/core/server-islands/vite-plugin-server-islands.js
import MagicString from "magic-string";
import { AstroError, AstroErrorData } from "../errors/index.js";
import { ASTRO_VITE_ENVIRONMENT_NAMES } from "../constants.js";
const SERVER_ISLAND_MANIFEST = "virtual:astro:server-island-manifest";
const RESOLVED_SERVER_ISLAND_MANIFEST = "\0" + SERVER_ISLAND_MANIFEST;
const serverIslandPlaceholderMap = "'$$server-islands-map$$'";
const serverIslandPlaceholderNameMap = "'$$server-islands-name-map$$'";
function vitePluginServerIslands({ settings }) {
  let command = "serve";
  let ssrEnvironment = null;
  const referenceIdMap = /* @__PURE__ */ new Map();
  const serverIslandMap = /* @__PURE__ */ new Map();
  const serverIslandNameMap = /* @__PURE__ */ new Map();
  return {
    name: "astro:server-islands",
    enforce: "post",
    config(_config, { command: _command }) {
      command = _command;
    },
    configureServer(server) {
      ssrEnvironment = server.environments[ASTRO_VITE_ENVIRONMENT_NAMES.ssr];
    },
    resolveId: {
      filter: {
        id: new RegExp(`^${SERVER_ISLAND_MANIFEST}$`)
      },
      handler() {
        return RESOLVED_SERVER_ISLAND_MANIFEST;
      }
    },
    load: {
      filter: {
        id: new RegExp(`^${RESOLVED_SERVER_ISLAND_MANIFEST}$`)
      },
      handler() {
        return {
          code: `export const serverIslandMap = ${serverIslandPlaceholderMap};

export const serverIslandNameMap = ${serverIslandPlaceholderNameMap};`
        };
      }
    },
    transform: {
      filter: {
        id: {
          include: [
            // Allows server islands in astro and mdx files
            /\.(astro|mdx)$/,
            new RegExp(`^${RESOLVED_SERVER_ISLAND_MANIFEST}$`)
          ]
        }
      },
      async handler(_code, id) {
        const info = this.getModuleInfo(id);
        const astro = info ? info.meta.astro : void 0;
        if (astro) {
          for (const comp of astro.serverComponents) {
            if (!serverIslandNameMap.has(comp.resolvedPath)) {
              if (!settings.adapter) {
                throw new AstroError(AstroErrorData.NoAdapterInstalledServerIslands);
              }
              let name = comp.localName;
              let idx = 1;
              while (true) {
                if (!serverIslandMap.has(name)) {
                  break;
                }
                name += idx++;
              }
              serverIslandNameMap.set(comp.resolvedPath, name);
              serverIslandMap.set(name, comp.resolvedPath);
              if (command === "build") {
                let referenceId = this.emitFile({
                  type: "chunk",
                  id: comp.specifier,
                  importer: id,
                  name: comp.localName
                });
                referenceIdMap.set(comp.resolvedPath, referenceId);
              }
            }
          }
        }
        if (serverIslandNameMap.size > 0 && serverIslandMap.size > 0 && ssrEnvironment) {
          const mod = ssrEnvironment.moduleGraph.getModuleById(RESOLVED_SERVER_ISLAND_MANIFEST);
          if (mod) {
            ssrEnvironment.moduleGraph.invalidateModule(mod);
          }
        }
        if (id === RESOLVED_SERVER_ISLAND_MANIFEST) {
          if (command === "build" && settings.buildOutput) {
            const hasServerIslands = serverIslandNameMap.size > 0;
            if (hasServerIslands && settings.buildOutput !== "server") {
              throw new AstroError(AstroErrorData.NoAdapterInstalledServerIslands);
            }
          }
          if (serverIslandNameMap.size > 0 && serverIslandMap.size > 0) {
            let mapSource = "new Map([\n	";
            for (let [name, path] of serverIslandMap) {
              mapSource += `
	['${name}', () => import('${path}')],`;
            }
            mapSource += "]);";
            return {
              code: `
					export const serverIslandMap = ${mapSource};
					

export const serverIslandNameMap = new Map(${JSON.stringify(Array.from(serverIslandNameMap.entries()), null, 2)});
					`
            };
          }
        }
      }
    },
    renderChunk(code, chunk) {
      if (code.includes(serverIslandPlaceholderMap)) {
        if (referenceIdMap.size === 0) {
          return {
            code: code.replace(serverIslandPlaceholderMap, "new Map();").replace(serverIslandPlaceholderNameMap, "new Map()"),
            map: null
          };
        }
        const isRelativeChunk = !chunk.isEntry;
        const dots = isRelativeChunk ? ".." : ".";
        let mapSource = "new Map([";
        let nameMapSource = "new Map(";
        for (let [resolvedPath, referenceId] of referenceIdMap) {
          const fileName = this.getFileName(referenceId);
          const islandName = serverIslandNameMap.get(resolvedPath);
          mapSource += `
	['${islandName}', () => import('${dots}/${fileName}')],`;
        }
        nameMapSource += `${JSON.stringify(Array.from(serverIslandNameMap.entries()), null, 2)}`;
        mapSource += "\n])";
        nameMapSource += "\n)";
        referenceIdMap.clear();
        const ms = new MagicString(code);
        ms.replace(serverIslandPlaceholderMap, mapSource);
        ms.replace(serverIslandPlaceholderNameMap, nameMapSource);
        return {
          code: ms.toString(),
          map: ms.generateMap({ hires: "boundary" })
        };
      }
    }
  };
}
export {
  SERVER_ISLAND_MANIFEST,
  vitePluginServerIslands
};
